connectivity/tests/service.go

```diff
@@ -16,128 +16,154 @@ package tests
-	for _, client := range c.ClientPods() {
-		serviceDestinations := serviceDefinitionMap{}
-		for _, echoSvc := range c.EchoServices() {
-			for _, echo := range c.EchoPods() {
-				if echo.Pod.Status.HostIP != client.Pod.Status.HostIP {
-					serviceDestinations[echo.Pod.Status.HostIP] = serviceDefinition{
-						port: int(echoSvc.Service.Spec.Ports[0].NodePort),
-						name: "NodePort",
-					}
+func (s *podToRemoteNodePort) Run(ctx context.Context, t *check.Test) {
+	var i int
+
+	for _, pod := range t.Context().ClientPods() {
+		for _, svc := range t.Context().EchoServices() {
+			for _, echo := range t.Context().EchoPods() {
+				// TODO(timo): Should probably use Cilium agent Pods or actual nodes as
+				// the innermost loop here, there is no actual correlation between
+				// echo Pods and nodes.
+				if pod.Pod.Status.HostIP != echo.Pod.Status.HostIP {
+					// If src and dst pod are running on different nodes,
+					// call the echo Pod's host IP on the service's NodePort.
+					curlNodePort(ctx, s, t, fmt.Sprintf("curl-%d", i), &pod, svc, &echo)
+
+					i++
 				}
 			}
 		}
-
-		testConnectivityToServiceDefinition(ctx, c, t, client, serviceDestinations)
 	}
 }
 
-type PodToLocalNodePort struct {
-	check.PolicyContext
-	Variant string
+// PodToLocalNodePort sends an HTTP request from all client Pods
+// to all echo Services' NodePorts, but only on the same node as
+// the client Pods.
+func PodToLocalNodePort(name string) check.Scenario {
+	return &podToLocalNodePort{
+		name: name,
+	}
 }
 
-func (t *PodToLocalNodePort) WithPolicy(yaml string) check.ConnectivityTest {
-	return t.WithPolicyRunner(t, yaml)
+// podToLocalNodePort implements a Scenario.
+type podToLocalNodePort struct {
+	name string
 }
 
-func (t *PodToLocalNodePort) Name() string {
-	return "pod-to-local-nodeport" + t.Variant
+func (s *podToLocalNodePort) Name() string {
+	tn := "pod-to-local-nodeport"
+	if s.name == "" {
+		return tn
+	}
+	return fmt.Sprintf("%s:%s", tn, s.name)
 }
 
-func (t *PodToLocalNodePort) Run(ctx context.Context, c check.TestContext) {
-	for _, client := range c.ClientPods() {
-		serviceDestinations := serviceDefinitionMap{}
-		for _, client := range c.ClientPods() {
-			for _, echoSvc := range c.EchoServices() {
-				for _, echo := range c.EchoPods() {
-					if echo.Pod.Status.HostIP == client.Pod.Status.HostIP {
-						serviceDestinations[echo.Pod.Status.HostIP] = serviceDefinition{
-							port: int(echoSvc.Service.Spec.Ports[0].NodePort),
-							name: "NodePort",
-						}
-					}
+func (s *podToLocalNodePort) Run(ctx context.Context, t *check.Test) {
+	var i int
+
+	for _, pod := range t.Context().ClientPods() {
+		for _, svc := range t.Context().EchoServices() {
+			for _, echo := range t.Context().EchoPods() {
+				// TODO(timo): Should probably use Cilium agent Pods or actual nodes as
+				// the innermost loop here, there is no actual correlation between
+				// echo Pods and nodes.
```

Comment:
Better to have a dedicated daemonset, eventually cilium pods should get stripped off unnecessary dependencies.

Reply:
There's already an internal list of Cilium pods, using that instead for now. Not sure if an extra DS would make sense, as Cilium is also a DS and expected to run everywhere. (at least within the context of the test suite)

An exception might be external workloads, but those are declared explicitly at the moment.